astroid/nodes/scoped_nodes/scoped_nodes.py

```diff
@@ -2810,7 +2811,9 @@ def implicit_metaclass(self):
             return builtin_lookup("type")[1][0]
         return None
 
-    def declared_metaclass(self, context=None):
+    def declared_metaclass(
+        self, context: InferenceContext | None = None
+    ) -> SuccessfulInferenceResult | None:
```

Comment:
Or do we want to assert the `infer()` to make sure we only return certain types? It doesn't make a lot of sense for a `metaclass` to be an `UnboundMethod` (for example), which it could be with the current typing and lack of assertions.

Reply:
From what I can tell or at least in the astroid tests. Only `nodes.ClassDef | None` is ever returned. I.e. it could actually make sense to restrict the return type here. Especially since `metaclass` could be called independently of `infer`.

I'm not sure how we can add an `assert` here though without changing the return generator expression. Might be better to use `type: ignore` or `cast` instead.

Reply:
My idea was to change the `next` call in the return statement to only allow ClassDef to pass through. That should work right?

Reply:
Ideally we should get a warning / error if somehow something else would be returned. Otherwise we might mask an issue that is entirely preventable.

I was also thinking about wether it should be `ClassDef` or `NodeNG`. Currently, I would lean towards `NodeNG`. It might not be as precise, but users can do a simple `isinstance` check afterwards. Or do you think `ClassDef` would be better?

Reply:
> Ideally we should get a warning / error if somehow something else would be returned. Otherwise we might mask an issue that is entirely preventable.

I don't really have any idea on how to allow this. Do you want to log something when this occurs?

> I was also thinking about wether it should be `ClassDef` or `NodeNG`. Currently, I would lean towards `NodeNG`. It might not be as precise, but users can do a simple `isinstance` check afterwards. Or do you think `ClassDef` would be better?

Yeah I think that's better. Especially considering plugins might also set meta classes and might not be as strict as we are.

Reply:
I have added the `NodeNG` check.